ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java

```diff
@@ -911,18 +968,31 @@ public void onDataNodeConfigChange(Iterable<DataNodeConfig> configs) {
     @Override
     public void onIdealStateChange(List<IdealState> idealState, NotificationContext changeContext) {
       if (!idealStateInitialized) {
-        logger.info("Received initial notification for IdealState change from {}", dcName);
+        logger.info("Received initial notification for IdealState change from {}", dcOrClusterName);
         idealStateInitialized = true;
       } else {
-        logger.info("IdealState change triggered from {}", dcName);
+        logger.info("IdealState change triggered from {}", dcOrClusterName);
       }
-      logger.debug("Detailed ideal states in {} are: {}", dcName, idealState);
+      logger.debug("Detailed ideal states in {} are: {}", dcOrClusterName, idealState);
       // rebuild the entire partition-to-resource map in current dc
       ConcurrentHashMap<String, String> partitionToResourceMap = new ConcurrentHashMap<>();
       for (IdealState state : idealState) {
         String resourceName = state.getResourceName();
         state.getPartitionSet().forEach(partitionName -> partitionToResourceMap.put(partitionName, resourceName));
       }
+      String dcName;
+      if (isAggregatedViewHandler) {
+        // Since helix aggregated view service doesn't aggregate ideal states across colos, we register ideal state
+        // change listener with helix service in each colo separately. So, this callback would be invoked for each
+        // data center separately. In order to know which data center, this callback corresponds to, we try to get one
+        // of the host names from Ideal state and get its data center name.
+        IdealState state = idealState.iterator().next();
+        String partition = state.getPartitionSet().iterator().next();
+        String instanceName = state.getInstanceSet(partition).iterator().next();
+        dcName = instanceNameToAmbryDataNode.get(instanceName).getDatacenterName();
```

Comment:
instanceNameToAmbryDataNode map is populated by DataNodeConfigSource, is there a way to guarantee the DataNodeConfigSource callback would be invoked before IdealState callback?

Reply:
I think we make sure that DataNodeConfigSource callback is invoked in the same context of thread that adds listener by [waiting](https://github.com/linkedin/ambry/blob/a6422967b7f01255d208da9f5c5cdb503451b257/ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java#L125) and fetching the information. Due to that, I think we can guarantee that `instanceNameToAmbryDataNode` is populated.

Reply:
that's right, that makes sense